Add support for bias in optimized op_linear.cpp.

hsharma35 opened the pull request:
Summary: Diff uses `op_add_sub_impl` to add bias after optimized gemm call.

Differential Revision: D75491158

swolchok commented on kernels/optimized/cpu/op_linear.cpp:
```diff
@@ -72,7 +70,13 @@ Tensor& opt_linear_out(
+  }
+
+  // Add bias.
+  return opt_add_sub_out_impl<false, op_name>(
```
Hmm. I don't have a great immediate intuition for whether this is better than the approach in #9527. Can you talk about (and perhaps add a comment explaining) why we shouldn't just rely on the accumulation in the GEMV/GEMM kernel? Possibly it just doesn't matter. Again CC @kimishpatel

hsharma35 replied:
Ah I didn't dig deeper into alternatives. This approach seemed to need the least code change. The approach in #9527 is likely faster. 
Can we push this PR in now and then let #9527 merge whenever possible? I just want bias support to unblock meta internal stuff.

kimishpatel replied:
Yeah I think #9527 is likely faster. If it is faster then why not just land that @hsharma35 ?

swolchok commented on shim_et/xplat/executorch/kernels/optimized/op_registration_util.bzl:
```diff
@@ -213,7 +213,11 @@ OPTIMIZED_ATEN_OPS = (
     op_target(
         name = "op_linear",
         deps = [
+            ":binary_ops",
+            ":add_sub_impl",
```
sort

swolchok commented on kernels/optimized/cpu/op_linear.cpp:
```diff
@@ -50,8 +47,9 @@ Tensor& opt_linear_out(
   }
+  static constexpr const char op_name[] = "linear.out";
   ET_SWITCH_REAL_TYPES_AND2(
-      Half, BFloat16, in.scalar_type(), ctx, "mm.out", CTYPE, [&]() {
```
Hmm, this is tricky. linear isn't a core ATen op so I  *think* we aren't supposed to do this; instead I *think* if there is a bias it's addmm.out and if not it's mm.out. This matters for selective build configurations and depends on how XNNPACK does it, so CC @kimishpatel / @digantdesai / @mcr229  ?

kimishpatel replied:
@larryliu0820 can you say something about how this impacts selective build? Are we relying on mm.out to apply selective dtype rule for all instances of op_mm?
@swolchok are you also saying that dtype selective build looks at op_mm in the edge ir graph and then apply selective build rule that match "mm.out" in the op name string?

larryliu0820 replied:
I think this is a right fix. For linear op we should use "linear.out" here. Selective build will have the following impact with this change: before this change, if a user select a dtype (say float16) for `mm.out`, it will apply to linear.out *by mistake*, like a side effect. After this change, any binary that relies this side effect will have a small binary size regression, because the key changes from "mm.out" to "linear.out". Though I would argue that this is still a right fix and we should fix whatever binary size regression by properly selecting "linear.out".

kimishpatel replied:
Thanks @larryliu0820 that was my understanding too. But @hsharma35 can we please use the other diff as the right fix. I dont think we should really land this as is if we have a better fix available. If this is urgent for you, lets prioritze the other one

kimishpatel reviewed (changes requested):
I want to block this for now in favor of landing the other fix for bias. mm.out -> linear.out is the right change though

kimishpatel commented on kernels/test/op_linear_test.cpp:
```diff
@@ -88,6 +96,70 @@ TEST_F(OpLinearOutTest, AllDtypesSupported) {
+  Tensor x = tf.full({kDimX, kReduceDim}, kValueX);
+  Tensor y = tf.full({kDimY, kReduceDim}, kValueY);
+  // Same size as output.
+  Tensor b = tf.full({kDimX, kDimY}, kValueBias);
```
I dont think your implementation supports this. In fact for linear bias cannot be 2d. Can you update test so taht expected values not the same for all elements. In that case you likely would have caught the issue. This test should otherwise actually fail

hsharma35 replied:
Good catch! I added this test to check if the aten implementation works. Surprisingly, the aten implementation can accept a 2d bias.
This works too:
```
x = torch.rand(3, 4)
w = torch.rand(5, 4)
b = torch.rand(3, 5)
out = torch.ops.aten.linear.default(x, w, b)
```
I've removed the 2D test. If we want to add support for 2D bias, it should be in a follow-up PR.

kimishpatel reviewed (changes requested):
ok so this diff does incorporate "idea" from the other diff so seems ok to me. but would like the test case to be updated

kimishpatel commented on kernels/optimized/cpu/op_linear.cpp:
```diff
@@ -6,30 +6,76 @@
+  // Output is a n x m x scalar_t, while bias is m x scalar_t.
+  const size_t row_size = static_cast<size_t>(m) * sizeof(scalar_t);
+  for (const auto col : c10::irange(n)) {
+    std::memcpy(
```
To handle 2d bias, you need to fix this. Bias pointer is not advancing

kimishpatel reviewed (approved):
I am accepting this but do note that you want either some random values or "not-all-outputs-are-same" values. The test as is, is quite weak IMO.

hsharma35 committed:
Add support for bias in optimized op_linear.cpp. (#11210)

Summary:

Diff initializes the output tensor before calling gemm with beta=1 when bias is non-nullopt.

Reviewed By: larryliu0820, zonglinpeng, kimishpatel

Differential Revision: D75491158
